common/quotas/global/algorithm/requestweighted.go

```diff
@@ -0,0 +1,442 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+/*
+Package algorithm contains a running-weighted-average calculator for ratelimits,
+and some associated types to prevent accidental confusion between the various
+floats/ints/etc involved.
+
+This package is intentionally unaware of RPC or any desired RPS limits, it just
+tracks the observed rates of requests and returns proportions, so other code can
+determine RPS values.
+
+This is both to simplify testing and to keep its internals as fast as possible, as
+it is relatively CPU heavy and needs to be quick to prevent a snowballing backlog
+of concurrent updates and requests for aggregated data.
+
+# How this fits in the global ratelimiting system
+
+Going back to the [github.com/uber/cadence/common/quotas/global] package diagram:
+each aggregating host will have one or more instances of this weight-calculator,
+and it will receive a shard worth of request data.
+
+Though this package has no direct connection to RPC structures, the request data is
+expected to closely match the argument to RequestWeighted.Update, and the response
+data like the return value of RequestWeighted.HostWeights.
+
+This makes aggregating hosts intentionally very simple outside of this package:
+they essentially just forward the request and response, multiplying by dynamicconfig
+intended-ratelimit values (which occurs outside the mutex, to minimize contention).
+
+# Expected use
+
+Limiting hosts collect metrics and submit it to an aggregating host via
+RequestWeighted.Update, through [some kind of RPC setup].
+
+Once updated, the aggregating host can get the RequestWeighted.HostWeights for
+that host's ratelimits (== the updated limits), multiply those 0..1 weights by
+the dynamicconfig configured RPS, and return them to the limiting host that
+triggered the request.
+
+If there are unused RPS remaining, the aggregating host *may* increase the RPS
+it returns by [some amount], to allow the limiting host to pre-emptively allow
+more requests than is "fair" before the next update.  Even if it does not, an
+increase in attempted usage will increase that limiter's weight on the next
+update cycle, so this is mostly intended as a tool for reducing incorrectly-rejected
+requests when a ratelimit's usage is well below its allowed limit.
+
+# Dealing with expired data
```

Comment:
yeah, the final goal for all this is to let us experiment internally while reusing the overall sharding / request flow / etc, since internally we have information that doesn't exist in other environments.

which is probably true of other environments too.  "customized" seems quite likely in general, though I do expect a majority will just have round-robin or random or something and won't need this.